misc/chattr.1
.\" -*- nroff -*-
.TH CHATTR 1 "March 1994" "Version 0.5"
.SH NAME
chattr \- change file attributes on a Linux second extended file system
.SH SYNOPSIS
.B chattr
[
.B \-RV
]
[
.B -v
version
]
[
mode
]
.I files...
.SH DESCRIPTION
.B chattr
changes the files attributes on an second extended file system.
.PP
The format of a symbolic mode is +-=[Scsu].
.PP
The operator `+' causes the selected attributes to be added to the
existing attributes of the files; `-' causes them to be removed; and
`=' causes them to be the only attributes that the files have.
.PP
The letters `Scsu' select the new attributes for the files: synchronous
updates (S), compressed (c), secure deletion (s), and undeletable (u).
.SH OPTIONS
.TP
.I -R
Recursively change attributes of directories and their contents.
.TP
.I -V
Verbosely describe changed attributes.
.TP
.I -v version
Set the files version.
.SH AUTHOR
.B chattr
has been written by Remy Card <card@masi.ibp.fr>, the developer and maintainer
of the ext2 fs.
.SH BUGS AND LIMITATIONS
As of ext2 fs 0.4, the kernel code only honours the `s' and `S' attributes. When
a file with the `s' attribute set is deleted, its blocks are zeroed and
written back to the disk. When a file with the `S' attribute set is modified,
the changes are written synchronously on the disk; this is equivalent to
the `sync' mount option applied to a subset of the files.
.PP
The other attributes will be implemented in a next ext2 fs version.
.SH AVAILABILITY
.B chattr
is available for anonymous ftp from ftp.ibp.fr (132.227.60.2) in
/pub/linux/BETA/ext2fs.
.SH SEE ALSO
.BR lsattr (1)


ChangeLog
Mon Oct 20 19:30:45 1997  Theodore Ts'o  <tytso@rsts-11.mit.edu>

	* configure.in (HAVE_LLSEEK_PROTOTYPE): Added check to see if
		llseek is declared in unistd.h

Sun Oct 19 19:09:30 1997  Theodore Ts'o  <tytso@rsts-11.mit.edu>

	* configure.in: Add tests for sys/stat.h, sys/time.h, and sys/types.h

Tue Jun 17 01:33:20 1997  Theodore Ts'o  <tytso@rsts-11.mit.edu>

	* Release of E2fsprogs 1.11

Sat Jun 14 03:26:45 1997  Theodore Ts'o  <tytso@rsts-11.mit.edu>

	* Makefile.in (SRCROOT): Allow people to set the version.h to
		something like 1.10-PLUS.

Sat Jun  7 16:38:40 1997  Theodore Ts'o  <tytso@rsts-11.mit.edu>

	* configure.in (rmakefile): Added (optional) private directory for
		resize2fs.

	* Makefile.in: Change recursive descent rules to check to see if a
		directory exists before trying to make it.

Thu May  8 22:23:49 1997  Theodore Ts'o  <tytso@rsts-11.mit.edu>

	* Makefile.in (source_tar_file): Fix up makefile to work in new
		CVS development environment.

Thu Apr 24 12:16:42 1997  Theodre Ts'o  <tytso@localhost.mit.edu>

	* Release of E2fsprogs version 1.10

Tue Apr 22 10:48:03 1997  Theodore Ts'o  <tytso@rsts-11.mit.edu>

	* configure.in: Add explicit check to see if linker accepts
		-static (since even Linux systems might not work if 
		libc.a isn't installed).

Thu Apr 17 12:23:38 1997  Theodore Ts'o  <tytso@rsts-11.mit.edu>

	* Release of E2fsprogs version 1.09

Fri Apr 11 18:56:26 1997  Theodore Ts'o  <tytso@rsts-11.mit.edu>

	* Release of E2fsprogs version 1.08

Wed Mar 12 13:32:05 1997  Theodore Y. Ts'o  <tytso@mit.edu>

	* Release of E2fsprogs version 1.07

Wed Jan 15 11:37:36 1997  Theodore Ts'o  <tytso@rsts-11.mit.edu>

	* config.sub (basic_machine): Added i686-* as another name for the
 		Pentium Pro.

Tue Oct  8 02:02:03 1996  Theodore Ts'o  <tytso@rsts-11.mit.edu>

	* Release of E2fsprogs version 1.06

Mon Oct  7 08:22:31 1996  Theodore Ts'o  <tytso@rsts-11.mit.edu>

	* Makefile.in (all): Don't run "make check" by default.  User
 	 	should manually run "make check" if they wish to test
 	 	things out.

Thu Sep 12 15:23:07 1996  Theodore Ts'o  <tytso@rsts-11.mit.edu>

	* Release of E2fsprogs version 1.05

Sat Aug 31 10:55:45 1996  Theodore Ts'o  <tytso@rsts-11.mit.edu>

	* configure.in (AC_CHECK_FUNCS): Add fchown to list of functions
	 	that we check.

Wed Aug 28 14:42:12 1996  Miles Bader  <miles@gnu.ai.mit.edu>

	* configure.in (usr_prefix): To be slightly more conformant with
		the coding standards, always default to ${prefix}
		unless on a linux system with prefix = ''.  Allow
		--with-usr-prefix option.

Tue Aug 27 16:53:29 1996  Miles Bader  <miles@gnu.ai.mit.edu>

	* configure.in (AC_CHECK_HEADERS): Add net/if.h & netinet/in.h.
		Add `--enable-fsck' switch, to allow configuration of
		fsck wrapper building (default yes except on the hurd).
		Make '' prefix default and LDFLAG_STATIC hacks work on
		the hurd as well as linux.

Tue Aug 27 16:23:56 1996  Theodore Ts'o  <tytso@rsts-11.mit.edu>

	* configure.in: Check to see if sys/types.h defines ino_t.  Add
		support for checking/sizing "long long".

Wed Aug 21 00:44:22 1996  Theodore Ts'o  <tytso@rsts-11.mit.edu>

	* configure.in: Added configure flag --enable-old-bitops, which
 		forces the bitops to use the standard bitmask operations.

Fri Aug  9 08:29:00 1996  Theodore Ts'o  <tytso@rsts-11.mit.edu>

	* configure.in: Check for existence of sys/utsname.h and
	 	strcasecmp().  Remove check for EXT2 fragment in system
	 	header file.  E2fsprogs now deals with the fragment fields
	 	by dispatching off of the OS field.

Tue Aug  6 14:34:19 1996  Theodore Ts'o  <tytso@rsts-11.mit.edu>

	* configure.in (AC_OUPUT): Create substitutions for the uuid
	 	library.

	* MCONFIG.in (all): Add new variables for the uuid library.

Thu May 23 12:39:07 1996  Theodore Ts'o  <tytso@rsts-11.mit.edu>

	* configure.in: Make the default prefix be '' for Linux.

Thu May 16 11:12:30 1996  Theodore Ts'o  <tytso@rsts-11.mit.edu>

	* Release of E2fsprogs version 1.04

Thu May 16 09:38:40 1996  Theodore Ts'o  <tytso@rsts-11.mit.edu>

	* configure.in: 
	  MCONFIG.in: Man pages, shell scripts and include files are now
		generated using the lib/substitute_sh script.  This is
		faster than the configure substitution, and it doesn't
		change the mod time of the file if it hasn't changed,
		which prevents needless compilation of files.

		Define new sets of Makefile variables: LIBSS, LIBCOM_ERR,
 		LIBEXT2FS, LIBE2P, and their static and profiled variants,
 		STATIC_* and PROFILED_*, which point to the actual file of
 		the shared or static library.  This way makefiles can link
 		directly with exactly the library they want.  Many ld's
		(include GNU ld) have a really broken idea of how -L
		works, and will link against an older library in /usr/lib
		even though there is a newer on in a specified -L directory.
		  
Wed May 15 21:40:22 1996  Theodore Ts'o  <tytso@rsts-11.mit.edu>

	* configure.in: Add support for --enable-dynamic-e2fsck, for
 		people who don't want to link e2fsck statically.  This
 		saves space, at the cost of increasing the reliance of
 		e2fsck other files (3-4 shared libraries).

Wed Mar 27 00:33:40 1996    <tytso@rsts-11.mit.edu>

	* Release of E2fsprogs version 1.03

Wed Jan 31 11:06:08 1996    <tytso@rsts-11.mit.edu>

	* Release of E2fsprogs version 1.02

Thu Oct 26 11:59:44 1995    <tytso@rsts-11.mit.edu>

	* configure.in (LDFALG_STATIC): Add temporary kludge for
		determining how to link executables statically.  For now,
		we assume that Linux systems only can link statically.

Fri Sep 22 18:23:55 1995    <tytso@rsts-11.mit.edu>

	* depfix.sed: Remove /usr/include lines which occur at the end of
		the dependency listing.

Mon Sep  4 21:40:22 1995  Remy Card  <card@bbj>

	* configure.in: Added support for the --enable-bsd-shlibs option.

Wed Aug  9 21:33:31 1995  Theodore Y. Ts'o  <tytso@dcl>

	* MCONFIG.in (LD): Add $(PURE) to the definition to make it easier
		to link executables using programs like purify.


misc/chattr.c
/*
 * chattr.c		- Change file attributes on an ext2 file system
 *
 * Copyright (C) 1993, 1994  Remy Card <card@masi.ibp.fr>
 *                           Laboratoire MASI, Institut Blaise Pascal
 *                           Universite Pierre et Marie Curie (Paris VI)
 *
 * This file can be redistributed under the terms of the GNU General
 * Public License
 */

/*
 * History:
 * 93/10/30	- Creation
 * 93/11/13	- Replace stat() calls by lstat() to avoid loops
 * 94/02/27	- Integrated in Ted's distribution
 */

#include <sys/types.h>
#include <dirent.h>
#include <fcntl.h>
#ifdef HAVE_GETOPT_H
#include <getopt.h>
#endif
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#ifdef HAVE_ERRNO_H
#include <errno.h>
#endif
#include <sys/param.h>
#include <sys/stat.h>
#include <linux/ext2_fs.h>

#include "et/com_err.h"
#include "e2p/e2p.h"

#include "../version.h"

const char * program_name = "chattr";

int add = 0;
int rem = 0;
int set = 0;
int set_version = 0;

unsigned long version;

int recursive = 0;
int verbose = 0;

unsigned long af;
unsigned long rf;
unsigned long sf;

static void volatile fatal_error (const char * fmt_string, int errcode)
{
	fprintf (stderr, fmt_string, program_name);
	exit (errcode);
}

#define usage() fatal_error ("usage: %s [-RV] [-+=acdisSu] [-v version] files...\n", \
			     1)

static int decode_arg (int * i, int argc, char ** argv)
{
	char * p;
	char * tmp;

	switch (argv[*i][0])
	{
	case '-':
		for (p = &argv[*i][1]; *p; p++)
			switch (*p)
			{
			case 'R':
				recursive = 1;
				break;
			case 'S':
				rf |= EXT2_SYNC_FL;
				rem = 1;
				break;
			case 'V':
				verbose = 1;
				break;
#ifdef	EXT2_APPEND_FL
			case 'a':
				rf |= EXT2_APPEND_FL;
				rem = 1;
				break;
#endif
			case 'c':
				rf |= EXT2_COMPR_FL;
				rem = 1;
				break;
#ifdef	EXT2_NODUMP_FL
			case 'd':
				rf |= EXT2_NODUMP_FL;
				rem = 1;
				break;
#endif
#ifdef	EXT2_IMMUTABLE_FL
			case 'i':
				rf |= EXT2_IMMUTABLE_FL;
				rem = 1;
				break;
#endif
			case 's':
				rf |= EXT2_SECRM_FL;
				rem = 1;
				break;
			case 'u':
				rf |= EXT2_UNRM_FL;
				rem = 1;
				break;
			case 'v':
				(*i)++;
				if (*i >= argc)
					usage ();
				version = strtol (argv[*i], &tmp, 0);
				if (*tmp)
				{
					com_err (program_name, 0,
						 "bad version - %s\n", argv[*i]);
					usage ();
				}
				set_version = 1;
				break;
			default:
				fprintf (stderr, "%s: Unrecognized argument: %c\n",
					 program_name, *p);
				usage ();
			}
		break;
	case '+':
		add = 1;
		for (p = &argv[*i][1]; *p; p++)
			switch (*p)
			{
			case 'S':
				af |= EXT2_SYNC_FL;
				break;
#ifdef	EXT2_APPEND_FL
			case 'a':
				af |= EXT2_APPEND_FL;
				break;
#endif
			case 'c':
				af |= EXT2_COMPR_FL;
				break;
#ifdef	EXT2_NODUMP_FL
			case 'd':
				af |= EXT2_NODUMP_FL;
				break;
#endif
#ifdef	EXT2_IMMUTABLE_FL
			case 'i':
				af |= EXT2_IMMUTABLE_FL;
				break;
#endif
			case 's':
				af |= EXT2_SECRM_FL;
				break;
			case 'u':
				af |= EXT2_UNRM_FL;
				break;
			default:
				usage ();
			}
		break;
	case '=':
		set = 1;
		for (p = &argv[*i][1]; *p; p++)
			switch (*p)
			{
			case 'S':
				sf |= EXT2_SYNC_FL;
				break;
#ifdef	EXT2_APPEND_FL
			case 'a':
				sf |= EXT2_APPEND_FL;
				break;
#endif
			case 'c':
				sf |= EXT2_COMPR_FL;
				break;
#ifdef	EXT2_NODUMP_FL
			case 'd':
				sf |= EXT2_NODUMP_FL;
				break;
#endif
#ifdef	EXT2_IMMUTABLE_FL
			case 'i':
				sf |= EXT2_IMMUTABLE_FL;
				break;
#endif
			case 's':
				sf |= EXT2_SECRM_FL;
				break;
			case 'u':
				sf |= EXT2_UNRM_FL;
				break;
			default:
				usage ();
			}
		break;
	default:
		return EOF;
		break;
	}
	return 1;
}

static int chattr_dir_proc (const char *, struct dirent *, void *);

static void change_attributes (const char * name)
{
	unsigned long flags;
	struct stat st;

	if (lstat (name, &st) == -1)
	{
		com_err (program_name, errno, "while stating %s", name);
		return;
	}
	if (set)
	{
		if (verbose)
		{
			printf ("Flags of %s set as ", name);
			print_flags (stdout, sf, 0);
			printf ("\n");
		}
		if (fsetflags (name, sf) == -1)
			perror (name);
	}
	else
	{
		if (fgetflags (name, &flags) == -1)
			com_err (program_name, errno,
			         "while reading flags on %s", name);
		else
		{
			if (rem)
				flags &= ~rf;
			if (add)
				flags |= af;
			if (verbose)
			{
				printf ("Flags of %s set as ", name);
				print_flags (stdout, flags, 0);
				printf ("\n");
			}
			if (fsetflags (name, flags) == -1)
				com_err (program_name, errno,
				         "while setting flags on %s", name);
		}
	}
	if (set_version)
	{
		if (verbose)
			printf ("Version of %s set as %lu\n", name, version);
		if (fsetversion (name, version) == -1)
			com_err (program_name, errno,
			         "while setting version on %s", name);
	}
	if (S_ISDIR(st.st_mode) && recursive)
		iterate_on_dir (name, chattr_dir_proc, (void *) NULL);
}

static int chattr_dir_proc (const char * dir_name, struct dirent * de, void * private)
{
	if (strcmp (de->d_name, ".") && strcmp (de->d_name, ".."))
	{
	        char *path;

		path = malloc(strlen (dir_name) + 1 + strlen (de->d_name) + 1);
		if (!path)
			fatal_error("Couldn't allocate path variable "
				    "in chattr_dir_proc", 1);
		sprintf (path, "%s/%s", dir_name, de->d_name);
		change_attributes (path);
		free(path);
	}
	return 0;
}

void main (int argc, char ** argv)
{
	int i, j;
	int end_arg = 0;

	fprintf (stderr, "chattr %s, %s for EXT2 FS %s, %s\n",
		 E2FSPROGS_VERSION, E2FSPROGS_DATE,
		 EXT2FS_VERSION, EXT2FS_DATE);
	if (argc && *argv)
		program_name = *argv;
	i = 1;
	while (i < argc && !end_arg)
	{
		if (decode_arg (&i, argc, argv) == EOF)
			end_arg = 1;
		else
			i++;
	}
	if (i >= argc)
		usage ();
	if (set && (add || rem))
	{
		fprintf (stderr, "= is incompatible with - and +\n");
		exit (1);
	}
	if (!(add || rem || set || set_version))
	{
		fprintf (stderr, "Must use '-v', =, - or +\n");
		exit (1);
	}
	for (j = i; j < argc; j++)
		change_attributes (argv[j]);
}
